### PR TITLE
Issue #3872: Add vertical tabs to user account form.

### LINKED ISSUE #3872: [UX] Clean up account edit page with vertical tabs
**Description of the need**
The current account/user edit page feels rather cluttered.

Screen Shot 2019-06-17 at 12 29 48 PM

**Proposed solution**
Would this be better with vertical tabs rather than a long list of fieldsets?

---

PR by @BWPanda: https://github.com/backdrop/backdrop/pull/2753

### PR DESCRIPTION
**Pull request by ghost**
Fixes https://github.com/backdrop/backdrop-issues/issues/3872

### CODE REVIEW

**Comment by laryn** [comment]
@BWPanda Can you close and open the PR? The sandbox doesn't seem to work.

File: core/modules/locale/locale.module [hunk 1]
```diff
@@ -199,11 +195,10 @@ function locale_form_alter(&$form, &$form_state, $form_id) {
     // Display language selector when either creating a user on the admin
     // interface or editing a user account.
     if ($form_id == 'user_register_form' || $form_id == 'user_profile_form') {
-      $selector = locale_language_selector_form($form['#user']);
+      locale_language_selector_form($form);
```
Comment (klonos) [inline core/modules/locale/locale.module]: I'm confused. Doesn't this function accept `$user` as its parameter? 🤔

Comment (ghost) [reply core/modules/locale/locale.module]: Yeah, I guess I changed it as I needed the full form to be sent so I could alter it... Can't recall now.

**Comment by herbdool** [comment]
Can't login so going to toggle the PR.

File: core/modules/contact/contact.module [hunk 2]
```diff
@@ -302,9 +302,10 @@ function contact_form_user_profile_form_alter(&$form, &$form_state) {
   $config = config('contact.settings');
   $form['contact'] = array(
     '#type' => 'fieldset',
-    '#title' => t('Contact settings'),
+    '#title' => t('Contact form'),
```
Comment (docwilmot) [inline core/modules/contact/contact.module:305]: Cant figure why this change is here.

File: core/modules/user/js/user.admin.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      return Backdrop.checkPlain(vals.join(', '));
+    });
+
+    // Personalization.
+    $context.find('fieldset#edit-personalization').backdropSetSummary(function() {
+      var vals = [];
+
+      // Signature.
+      var signature = $context.find('textarea[name="signature[value]"]');
+      console.log(signature);
```
Comment (docwilmot) [inline core/modules/user/js/user.admin.js]: Stray

File: core/modules/user/js/user.admin.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      }
+
+      // Picture.
+      var pictureNew = $context.find('input[name="files[picture_upload]"]');
+      var pictureExisting = $context.find('.user-picture');
+      if ((pictureNew.length && pictureNew.val().length) || pictureExisting.length) {
+        vals.push(Backdrop.t('Picture'));
+      }
+
+      return Backdrop.checkPlain(vals.join(', '));
```
Comment (docwilmot) [inline core/modules/user/js/user.admin.js:62]: Could this use a default summary string if `vals` is empty? No summary seems off.

File: core/modules/user/js/user.admin.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    });
+
+    // Personalization.
+    $context.find('fieldset#edit-personalization').backdropSetSummary(function() {
+      var vals = [];
+
+      // Signature.
+      var signature = $context.find('textarea[name="signature[value]"]');
```
Comment (docwilmot) [inline core/modules/user/js/user.admin.js]: Extremely nitpicky and unnecessary comment follows, and core is quite variable with this, but few other examples in this file where `find()` returns a JQuery object, and convention suggests we name the variable with the `$`

**Review by docwilmot — COMMENTED** [review]
LGTM, minor comments.

File: core/modules/contact/contact.module (hunk 2 above)
Comment (quicksketch) [reply core/modules/contact/contact.module:305]: The change seems optional. I'm guessing it was changed because the only setting within this fieldset was whether to enable the "personal contact form".

File: core/modules/user/user.module [hunk 6]
```diff
@@ -936,49 +944,57 @@ function user_account_form(&$form, &$form_state) {
...
-    '#title' => t('Delete picture'),
-    '#access' => !empty($account->picture->fid),
-    '#description' => t('Check this box to delete your current picture.'),
-  );
-  $form['picture']['picture_upload'] = array(
-    '#type' => 'file',
-    '#title' => t('Upload picture'),
-    '#size' => 48,
-    '#description' => t('Your virtual face or picture. Pictures larger than @dimensions pixels will be scaled down.', array('@dimensions' => $site_config->get('user_picture_dimensions'))) . ' ' . filter_xss_admin($site_config->get('user_picture_guidelines')),
+    '#title' => t('Regional and language'),
```
Comment (quicksketch) [inline core/modules/user/user.module]: Can we change this heading (and the form key `regional_language`) to "Region and Language" (`region_language`)? That would match terminology used by OSes

MacOS:
image

Windows:
image

iPhone:
image

Ubuntu:
image


The key `regional_language` led me to believe it was for dialects specific to a particular region (a "regional language").

File: core/modules/user/js/user.admin.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      var vals = [];
+
+      if ($context.find('input[name="contact"]:checked').length) {
+        vals.push(Backdrop.t('Enabled'));
+      }
+      else {
+        vals.push(Backdrop.t('Disabled'));
+      }
+
+      return Backdrop.checkPlain(vals.join(', '));
```
Comment (quicksketch) [inline core/modules/user/js/user.admin.js]: No need to concatenate values if we're only returning a single word:

```suggestion
      if ($context.find('input[name="contact"]:checked').length) {
        return Backdrop.t('Enabled');
      }
      else {
        return Backdrop.t('Disabled');
      }
```

File: core/modules/user/js/user.admin.js (hunk 3 above)
Comment (herbdool) [inline core/modules/user/js/user.admin.js]:
```suggestion
```

File: core/modules/user/js/user.admin.js (hunk 5 above)
Comment (herbdool) [inline core/modules/user/js/user.admin.js]:
```suggestion
      var $signature = $context.find('textarea[name="signature[value]"]');
```

File: core/modules/user/js/user.admin.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    });
+
+    // Personalization.
+    $context.find('fieldset#edit-personalization').backdropSetSummary(function() {
+      var vals = [];
+
+      // Signature.
+      var signature = $context.find('textarea[name="signature[value]"]');
+      console.log(signature);
+      if (signature.length && signature.val().length) {
```
Comment (herbdool) [inline core/modules/user/js/user.admin.js]:
```suggestion
      if ($signature.length && $signature.val().length) {
```

File: core/modules/user/js/user.admin.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+      var vals = [];
+
+      // Status.
+      var status = $context.find('input[name="status"]:checked').parent().find('label').text();
+      vals.push($.trim(status));
+
+      // Roles.
+      var roles = [];
+      var checkedRoles = $context.find('input[name^="roles"]:not([disabled]):checked');
+      checkedRoles.each(function() {
```
Comment (herbdool) [inline core/modules/user/js/user.admin.js]:
```suggestion
      var $checkedRoles = $context.find('input[name^="roles"]:not([disabled]):checked');
      $checkedRoles.each(function() {
```

File: core/modules/user/js/user.admin.js [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      var signature = $context.find('textarea[name="signature[value]"]');
+      console.log(signature);
+      if (signature.length && signature.val().length) {
+        vals.push(Backdrop.t('Signature'));
+      }
+
+      // Picture.
+      var pictureNew = $context.find('input[name="files[picture_upload]"]');
+      var pictureExisting = $context.find('.user-picture');
+      if ((pictureNew.length && pictureNew.val().length) || pictureExisting.length) {
```
Comment (herbdool) [inline core/modules/user/js/user.admin.js]:
```suggestion
      var $pictureNew = $context.find('input[name="files[picture_upload]"]');
      var $pictureExisting = $context.find('.user-picture');
      if (($pictureNew.length && $pictureNew.val().length) || $pictureExisting.length) {
```

File: core/modules/user/js/user.admin.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+    });
+
+    // Region and language.
+    $context.find('fieldset#edit-regional-language').backdropSetSummary(function() {
+      var vals = [];
+
+      // Timezone.
+      var timezone = $context.find('select[name="timezone"]');
+      if (timezone.length && timezone.val().length) {
+        vals.push($.trim(timezone.val()));
```
Comment (herbdool) [inline core/modules/user/js/user.admin.js]:
```suggestion
      var $timezone = $context.find('select[name="timezone"]');
      if ($timezone.length && $timezone.val().length) {
        vals.push($.trim($timezone.val()));
```

File: core/modules/user/js/user.admin.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+      }
+
+      // Language.
+      var language = $context.find('input[name="language"]:checked');
+      if (language.length) {
+        vals.push($.trim(language.parent().find('label').text()));
```
Comment (herbdool) [inline core/modules/user/js/user.admin.js]:
```suggestion
      var $language = $context.find('input[name="language"]:checked');
      if ($language.length) {
        vals.push($.trim($language.parent().find('label').text()));
```

File: core/modules/user/user.module (hunk 6 above)
Comment (herbdool) [reply core/modules/user/user.module]:
```suggestion
    '#title' => t('Region and language'),
```

File: core/modules/user/user.module [hunk 13]
```diff
@@ -936,49 +944,57 @@ function user_account_form(&$form, &$form_state) {
...
+    $form['personalization']['picture_upload'] = array(
+      '#type' => 'file',
+      '#title' => t('Upload picture'),
+      '#size' => 48,
+      '#description' => t('Your virtual face or picture. Pictures larger than @dimensions pixels will be scaled down.', array('@dimensions' => $site_config->get('user_picture_dimensions'))) . ' ' . filter_xss_admin($site_config->get('user_picture_guidelines')),
+    );
+    $form['#validate'][] = 'user_validate_picture';
+  }
+
+  $form['regional_language'] = array(
```
Comment (herbdool) [inline core/modules/user/user.module]:
```suggestion
  $form['region_language'] = array(
```

File: core/modules/system/system.module [hunk 14]
```diff
@@ -2482,21 +2482,15 @@ function system_user_timezone(&$form, &$form_state) {
...
+  $form['regional_language']['timezone'] = array(
     '#type' => 'select',
     '#title' => t('Time zone'),
     '#default_value' => isset($account->timezone) ? $account->timezone : ($account->uid == $user->uid ? config_get('system.date', 'default_timezone') : ''),
     '#options' => system_time_zones($account->uid != $user->uid),
     '#description' => t('Select the desired local time and time zone. Dates and times throughout this site will be displayed using this time zone.'),
   );
   if (!isset($account->timezone) && $account->uid == $user->uid && empty($form_state['input']['timezone'])) {
-    $form['timezone']['timezone']['#attributes'] = array('class' => array('timezone-detect'));
+    $form['regional_language']['timezone']['#attributes'] = array('class' => array('timezone-detect'));
```
Comment (herbdool) [inline core/modules/system/system.module]:
```suggestion
    $form['region_language']['timezone']['#attributes'] = array('class' => array('timezone-detect'));
```

File: core/modules/system/system.module [hunk 15]
```diff
@@ -2482,21 +2482,15 @@ function system_user_timezone(&$form, &$form_state) {
...
 
   $account = $form['#user'];
-  $form['timezone'] = array(
-    '#type' => 'fieldset',
-    '#title' => t('Locale settings'),
-    '#weight' => 6,
-    '#collapsible' => TRUE,
-  );
-  $form['timezone']['timezone'] = array(
+  $form['regional_language']['timezone'] = array(
```
Comment (herbdool) [inline core/modules/system/system.module]:
```suggestion
  $form['region_language']['timezone'] = array(
```

File: core/modules/locale/locale.module [hunk 16]
```diff
@@ -201,10 +201,11 @@ function locale_form_alter(&$form, &$form_state, $form_id) {
     // interface or editing a user account.
     if ($form_id == 'user_register_form' || $form_id == 'user_profile_form') {
       $selector = locale_language_selector_form($form['#user']);
+      $selector['locale']['#type'] = 'container';
+      $form['regional_language'] += $selector;
       if ($form_id == 'user_register_form') {
-        $selector['locale']['#access'] = user_access('administer users');
+        $form['regional_language']['locale']['#access'] = user_access('administer users');
```
Comment (herbdool) [inline core/modules/locale/locale.module]:
```suggestion
        $form['region_language']['locale']['#access'] = user_access('administer users');
```

File: core/modules/locale/locale.module [hunk 17]
```diff
@@ -201,10 +201,11 @@ function locale_form_alter(&$form, &$form_state, $form_id) {
     // interface or editing a user account.
     if ($form_id == 'user_register_form' || $form_id == 'user_profile_form') {
       $selector = locale_language_selector_form($form['#user']);
+      $selector['locale']['#type'] = 'container';
+      $form['regional_language'] += $selector;
```
Comment (herbdool) [inline core/modules/locale/locale.module]:
```suggestion
      $form['region_language'] += $selector;
```

File: core/modules/user/js/user.admin.js [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+    // Personalization.
+    $context.find('fieldset#edit-personalization').backdropSetSummary(function() {
+      var vals = [];
+
+      // Signature.
+      var signature = $context.find('textarea[name="signature[value]"]');
+      console.log(signature);
+      if (signature.length && signature.val().length) {
+        vals.push(Backdrop.t('Signature'));
+      }
```
Comment (herbdool) [inline core/modules/user/js/user.admin.js]:
```suggestion
        var signatureText = Backdrop.t('Signature');
      }
      else {
        var signatureText = Backdrop.t('No signature');
      }
      vals.push(signatureText);
```

File: core/modules/user/js/user.admin.js [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+
+      // Picture.
+      var pictureNew = $context.find('input[name="files[picture_upload]"]');
+      var pictureExisting = $context.find('.user-picture');
+      if ((pictureNew.length && pictureNew.val().length) || pictureExisting.length) {
+        vals.push(Backdrop.t('Picture'));
+      }
```
Comment (herbdool) [inline core/modules/user/js/user.admin.js]:
```suggestion
        var pictureText = Backdrop.t('Picture');
      }
      else {
        var pictureText = Backdrop.t('No picture');
      }
      vals.push(pictureText);
```

**Review by herbdool — CHANGES_REQUESTED** [review]
I went through and made other reviewers suggestions into commit suggestions so they're easier to merge in if wanted.

**Review by herbdool — APPROVED** [review]

**Comment by quicksketch** [comment]
Merged into 1.x for 1.23.0 (this will be in the preview release)! Thank you!
